ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go

```diff
@@ -0,0 +1,169 @@
+	ProtocolRule *IngressNodeFirewallProtoRule `json:"protoRule,omitempty"`
+
+	// ICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols.
+	// +kubebuilder:validation:Optional
+	ICMPRule *IngressNodeFirewallICMPRule `json:"icmpRule,omitempty"`
+}
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+	// Order define order of execution of ingress firewall rules .
+	// + index 0 is used internally as catch all for unclassified packets matching the same sourceCIDR.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum:=1
+	Order uint32 `json:"order"`
+
+	// ProtocolConfig is a discriminated union of protocol's specific configuration.
+	// if not specificed matching will be based on protocol value and the protocol configs like dstPort/type/code will be ignored
+	// +kubebuilder:validation:Optional
+	ProtocolConfig IngressNodeProtocolConfig `json:"protocolConfig"`
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol"`
+
+	// Action can be Allow or Deny, default action is deny
+	// +kubebuilder:validation:Optional
+	Action IngressNodeFirewallActionType `json:"action,omitempty"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type IngressNodeFirewallRuleProtocolType string
+
+const (
+	// ProtocolTypeICMP refers to the ICMP protocol
+	ProtocolTypeICMP IngressNodeFirewallRuleProtocolType = "ICMP"
+
+	// ProtocolTypeICMP6 refers to the ICMPv6 protocol
+	ProtocolTypeICMP6 IngressNodeFirewallRuleProtocolType = "ICMPv6"
+
+	// ProtocolTypeTCP refers to the TCP protocol, for either IPv4 or IPv6
+	ProtocolTypeTCP IngressNodeFirewallRuleProtocolType = "TCP"
+
+	// ProtocolTypeUDP refers to the UDP protocol, for either IPv4 or IPv6
+	ProtocolTypeUDP IngressNodeFirewallRuleProtocolType = "UDP"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol, for either IPv4 or IPv6
+	ProtocolTypeSCTP IngressNodeFirewallRuleProtocolType = "SCTP"
+)
+
+// IngressNodeFirewallActionType indicates whether an IngressNodeFirewallRule allows or denies traffic
+// +kubebuilder:validation:Enum="Allow";"Deny"
+type IngressNodeFirewallActionType string
+
+const (
+	IngressNodeFirewallAllow IngressNodeFirewallActionType = "Allow"
+	IngressNodeFirewallDeny  IngressNodeFirewallActionType = "Deny"
+)
+
+// IngressNodeFirewallRules define ingress node firewall rule
+type IngressNodeFirewallRules struct {
+	// SourceCIDRS is A list of CIDRs from which we apply node firewall rule
+	// +kubebuilder:validation:one-of[]=Type=string;Format:cidr
+	// +kubebuilder:validation:MinItems:=1
+	SourceCIDRs []string `json:"sourceCIDRs"`
+	// FirewallProtocolRules is A list of per protocol ingress node firewall rules
+	// +kubebuilder:listType:=map
+	// +kubebuilder:listMapKey:=order
```

Comment:
These don't use the the kubebuilder format
```suggestion
	// +listType=map
	// +listMapKey=order
```